### ISSUE TITLE
Custom VJP involving BCOO producing incorrect gradients (dependent on transpose)

### ISSUE THREAD

**Issue by alecjacobson** [jax-ml/jax#12811]
### Description

Here's a [collab](https://colab.research.google.com/drive/1sKBUK-7I7eEBRp94iv-iF4Gpbvfe1rT6?usp=sharing#scrollTo=0hZBBas8E-Vf) to reproduce the issue. I'll repeat it here with annotation.

I'd like to write a custom_vjp for a function involving a BCOO sparse matrix. I've minified my example to consider the toy case of matrix-vector multiplication.

```python
import jax.numpy as jnp
from jax.experimental.sparse import BCOO 
from jax import grad, jit, vmap
from jax import custom_jvp, custom_vjp

# variable parameter
p = jnp.array(1.0)
# constants
b = jnp.array([2.0,3.0])
x = jnp.array([-3.0,2.0])

#############################################################
# Dense version without custom function
#############################################################

# E(p) = ‖ A(p) b - x ‖²
def E_dense(p):
    a = jnp.array([jnp.cos(p),-jnp.sin(p),jnp.sin(p),jnp.cos(p)])
    # build a dense matrix that depense on p
    I = jnp.array([0,1,0,1])
    J = jnp.array([0,0,1,1])
    A = jnp.zeros((2,2)).at[(I,J)].set(a)
    return jnp.linalg.norm(A@b - x)**2

print("E_dense(p): %g" % E_dense(p))
print("∇E_dense(p): %g" % grad(E_dense)(p))
print("----------------------------------------------------")

#############################################################
# Sparse version without custom_vjp
#############################################################

def E_sparse(p):
    a = jnp.array([jnp.cos(p),-jnp.sin(p),jnp.sin(p),jnp.cos(p)])
    I = jnp.array([0,1,0,1])
    J = jnp.array([0,0,1,1])
    A = BCOO((a,jnp.vstack((I,J)).T), shape=(2,2))
    return jnp.linalg.norm(A@b - x)**2

print("E_sparse(p): %g" % E_sparse(p))
print("∇E_sparse(p): %g" % grad(E_sparse)(p))
print("----------------------------------------------------")


#############################################################
# Dense version using custom_vjp
#############################################################

@custom_vjp
def mm_dense(A,b):
    return A @ b

def mm_dense_fwd(A,b):
    return mm_dense(A,b), (A,b)

def mm_dense_bwd(res,g):
    A,b = res
    partial1 = (jnp.outer(g , b))
    partial2 = (A.T  @ g)
    print("[mm_dense_bwd] partial1:")
    print(partial1)
    return ( partial1 , partial2 )

mm_dense.defvjp(mm_dense_fwd, mm_dense_bwd)

def E_dense_mm(p):
    a = jnp.array([jnp.cos(p),-jnp.sin(p),jnp.sin(p),jnp.cos(p)])
    I = jnp.array([0,1,0,1])
    J = jnp.array([0,0,1,1])
    A = jnp.zeros((2,2)).at[(I,J)].set(a,unique_indices=True)
    return jnp.linalg.norm(mm_dense(A,b) - x)**2

print("E_dense_mm(p): %g" % E_dense_mm(p))
print("∇E_dense_mm(p): %g" % grad(E_dense_mm)(p))
print("----------------------------------------------------")

#############################################################
# Sparse version using custom_vjp
#############################################################

# Given a vector, construct a sparse diagonal matrix
def diag_BCOO(b):
    indices = jnp.vstack((jnp.arange(0,len(b)),jnp.arange(0,len(b)))).T
    return BCOO( (b, indices), shape=(len(b),len(b)))

@custom_vjp
def mm_sparse(A,b):
    return A @ b

def mm_sparse_fwd(A,b):
    # cache the sparsity pattern of A
    S = BCOO((jnp.ones_like(A.data),A.indices),shape=A.shape)
    return mm_sparse(A,b), (A,b,S)

def mm_sparse_bwd(res,g):
    A,b,S = res
    partial1 = diag_BCOO(g) @ S @ diag_BCOO(b)
    # 🥝 This works even though it's returning the wrongly transposed partial
    #partial1 = diag_BCOO(b) @ S.T @ diag_BCOO(g)
    # 🍌 This also works, so it seems the final .T doesn't change anything
    #partial1 = (diag_BCOO(b) @ S.T @ diag_BCOO(g)).T
    print("[mm_sparse_bwd] partial1:")
    print(partial1.todense())
    partial2 = (A.T  @ g)
    return (partial1 , partial2)

mm_sparse.defvjp(mm_sparse_fwd, mm_sparse_bwd)

# E(p) = ‖ A(p) b - x ‖²
def E_sparse_mm(p):
    a = jnp.array([jnp.cos(p),-jnp.sin(p),jnp.sin(p),jnp.cos(p)])
    I = jnp.array([0,1,0,1])
    J = jnp.array([0,0,1,1])
    A = BCOO((a,jnp.vstack((I,J)).T), shape=(2,2))
    return jnp.linalg.norm(mm_sparse(A,b) - x)**2

print("E_sparse_mm(p): %g" % E_sparse_mm(p))
print("∇E_sparse_mm(p): %g" % grad(E_sparse_mm)(p))
```

Running this will output:

```
E_dense(p): 47.8782
∇E_dense(p): 14.0479
----------------------------------------------------
E_sparse(p): 47.8782
∇E_sparse(p): 14.0479
----------------------------------------------------
E_dense_mm(p): 47.8782
[mm_dense_bwd] partial1:
[[ 26.42007    39.630104 ]
 [ -8.24814   -12.3722105]]
∇E_dense_mm(p): 14.0479
----------------------------------------------------
E_sparse_mm(p): 47.8782
[mm_sparse_bwd] partial1:
[[ 26.42007    39.630104 ]
 [ -8.24814   -12.3722105]]
∇E_sparse_mm(p): -37.6896
```

We can see that the `∇E_sparse_mm(p): -37.6896` is incorrect, yet the `[mm_sparse_bwd] partial1:` is computed to be the same as the dense version.

It gets very interesting if you start commenting out the 🥝 and 🍌 lines. The 🥝 line computes the _**wrongly**_ transposed `partial1` yet results in the correct `∇E_sparse_mm`:

```
[mm_sparse_bwd] partial1:
[[ 26.42007    -8.24814  ]
 [ 39.630104  -12.3722105]]
∇E_sparse_mm(p): 14.0479
```

Adding another transpose `.T` (🍌 ) doesn't seem to have an effect on the `∇E_sparse_mm` but it flips the `partial1` back to being correct.

```
[mm_sparse_bwd] partial1:
[[ 26.42007    39.630104 ]
 [ -8.24814   -12.3722105]]
∇E_sparse_mm(p): 14.0479
```

I believe this is a bug within Jax (but would be happy to find out I'm doing something wrong/unexpected).

> FWIW, I have a guess that this might be due to transposes being tracked with a Bool flag and not being read correctly during the gradient tracing. I have no basis for this but just a hunch based on working with other sparse matrix libraries.

### What jax/jaxlib version are you using?

jaxv0.3.14

### Which accelerator(s) are you using?

CPU

### Additional system info

python 3.8 on mac os 12.6 M1

**Comment by jakevdp** [comment]
Thanks for the detailed report – I don't have any insight into what's going on, but I will say that to my knowledge this is the first time anyone has tried using BCOO objects within a custom_jvp. I think it's very possible that there's some fundamental incompatiblity between the two that we haven't yet thought of.

**Comment by jakevdp** [comment]
I'm going to close this, because we don't have plans to further develop `jax.experimental.sparse`
